### DISCUSSION TITLE
Port highlight

### DISCUSSION THREAD

**Discussion by snehabihani** [Blazor-Diagrams/Blazor.Diagrams#373] [Q&A]
Hi if i have link between two ports when i click link it is getting slected can we highlight the ports also which are connected by the link? I have many ports on for each node. it is very difficult to figure out.
please help

**Comment by K0369** [comment]
You can add classes to the ports html element by supplying them to the port-renderer (as described in the documentation: [https://blazor-diagrams.zhaytam.com/documentation/ports](url)

Here is a little bit of code that should get you going:

<details><summary>1. Create a custom port model that holds the information if it is selected or not:</summary>
<p>

```c#
    public class CustomPort : PortModel
    {
        public CustomPort(NodeModel parent, PortAlignment alignment = PortAlignment.Bottom, Point? position = null, Size? size = null) : base(parent, alignment, position, size)
        {
        }

        public CustomPort(string id, NodeModel parent, PortAlignment alignment = PortAlignment.Bottom, Point? position = null, Size? size = null) : base(id, parent, alignment, position, size)
        {
        }

        public bool Highlighted { get; set; } = false;

    }
``` 

</p>
</details> 



 <details><summary>2. Create a custom node widget, that adds the "highlighted" information to the port-renderer class attribute</summary>
<p>

```c#
<div class="default-node @(Node.Group != null ? "grouped" : "") @(Node.Selected ? "selected" : "")">
    @(Node.Title ?? "Title")
    @foreach (var port in Node.Ports)
    {
        @if (port is CustomPort custom)
        {
            <Blazor.Diagrams.Components.Renderers.PortRenderer @key="port" Port="port"
                                                               Class="@GetClass(port)">
            </Blazor.Diagrams.Components.Renderers.PortRenderer>
        }
        else
        {
            <Blazor.Diagrams.Components.Renderers.PortRenderer @key="port" Port="port" Class="default"></Blazor.Diagrams.Components.Renderers.PortRenderer>
        }
    }
</div>

@code {

    [Parameter]
    public NodeModel Node { get; set; } = null!;

    private string GetClass(PortModel port)
    {
        var builder = new StringBuilder("default");
        if (port is CustomPort custom)
        {
            builder.Append($"{(custom.Highlighted ? " highlighted" : "")}");

        }
        return builder.ToString();
    }
}
``` 

</p>
</details> 


 <details><summary>3. Create a method that sets the flag in the custom port model:</summary>
<p>

```c#
private void HighlightPort(SelectableModel model)
{
    if (model is not BaseLinkModel link) return;
    SetPortHighlights(link.Source, link.Selected);
    SetPortHighlights(link.Target, link.Selected);

}

private void SetPortHighlights(Anchor anchor, bool targetState)
{
    if (anchor is SinglePortAnchor singlePortAnchor
        && singlePortAnchor.Port is CustomPort port)
    {
        port.Highlighted = targetState;
        port.Parent.RefreshAll();
        port.Refresh();
    }
}
``` 

</p>
</details> 


<details><summary>4. Register the custom node component and the method on the diagram</summary>
<p>

```c#
Diagram.RegisterComponent<NodeModel, CustomNodeWidget>();
Diagram.SelectionChanged += HighlightPort;
``` 

</p>
</details> 

<details><summary>5. Add some .css magic to highlight the port</summary>
<p>

```css
.diagram-node .default-node .diagram-port.highlighted {
    background-color: darkorange;
}
``` 

</p>
</details> 



<details><summary>Result:</summary>

Not selected:
![image](https://github.com/Blazor-Diagrams/Blazor.Diagrams/assets/41567572/e2eb0ce9-3e7a-4c8d-a95f-65700c6983d7)

Left link selected:
![image](https://github.com/Blazor-Diagrams/Blazor.Diagrams/assets/41567572/64710744-9a0b-42ca-948d-770ccc160e55)

Right link selected:
![image](https://github.com/Blazor-Diagrams/Blazor.Diagrams/assets/41567572/20819720-2426-4c91-a1ee-d8b6e680f987)

**Reply by snehabihani** [reply]
Thank you all for the quick soltion

**Comment by snehabihani** [comment]
Thanks for quick help.
